Take args from JSON post data

abrandemuehl opened the pull request:
Makes a change in the API to allow for multiple locations to be passed

Resolves #7

abrandemuehl committed:
Take args from JSON post data

Makes a change in the API to allow for multiple locations to be passed

Resolves #7

abrandemuehl commented:
The post data should look like:

```json
{"items": ["A1", "B2", "C3"]}
```

sameetandpotatoes commented:
@abrandemuehl So would this require the merch service to make one request to `/vend` with all items? Should this also return an array of successes/failures so that I know which transactions specifically succeeded?

abrandemuehl commented:
Right now, if it fails in the middle, then problems will happen. I'll update it to catch failures and return which drinks were successfully vended

abrandemuehl commented:
How would you prefer success/failure of individual vends be indicated? I can return a combination of the index of what failed and the location.

Also, if the vend_queue feature gets put through, it'll make vending asynchronous with respect to the api requests made. We could either have the async vend queue notify the merch service at the end of the vend, or we could ditch the async vend queue for the for loop in this PR. They both work, but this one will be simpler

abrandemuehl committed:
Add vending error checking

The machine will return the index of the vend that failed

sameetandpotatoes commented:
Hmm, well ideally I was thinking something like this:

I send:
```
{
    "transaction_id": 1,
    "items": [ "A1", "B2", "C3" ]
}
```

and get back:
```
{
   "transaction_id": 1,
   "items": [
       {
             "error": null,
             "location": "A1"
       }, {
             "error": "No item present",
             "location": "B2",
       }, {
             "error": null,
             "location": "C3"
       }
   ]
}
```

It would be nice to get that same transaction id back so with a vend queue it's clear what transaction  these items belonged to (as transaction id comes from the merch service's database).

I'm open to suggestions though.

abrandemuehl committed:
Add success or failure to vend endpoint

abrandemuehl commented:
Updated

abrandemuehl committed:
Update README with API documentation

abrandemuehl commented:
I've added the documentation for the API to the readme

abrandemuehl commented:
What do you think about the final result @sameetandpotatoes. I'll get this tested on the machine sometime this week.

narendasan commented on README.md:
```diff
@@ -1,9 +1,46 @@
+Every request through the API has a token in the header that is unique to groot.
+This way, only groot can make requests to vend things.
+The token is stored in the request header.
+
```
Should be something like 

"Requests to the device must contain a valid token in the Authorization Header for a request to be processed. 
As of now the only token will be given solely to the groot merch service, so if you wish to make merch requests go through groot. "

Also rename the header field from Token to Authorization, better match with web conventions (may require change by @sameetandpotatoes on the service side)

sameetandpotatoes commented:
Yeah this looks good to me. Before this merges though, I will need to make the corresponding change in `groot-merch-service` and then we should probably test that too.

rauhul reviewed (changes requested):
InvalidLocationError should be thrown for any location the machine physically cannot vend. If, for some reason, Groot ever needs to press another key we can create a different endpoint to do so.

abrandemuehl committed:
Raise InvalidLocationError on Invalid Locations

abrandemuehl committed:
Fix bounds checking

rauhul reviewed (approved):
Seems good to the guy that doesn't know python 👍

rauhul commented:
@sameetandpotatoes can you make those changes to the service

narendasan commented on machine_controller/app.py:
```diff
@@ -45,12 +45,28 @@
 def hello_world():
     if request.headers.get('TOKEN', '') != token_value:
```
Change this from ```TOKEN``` to ```Authorization```

narendasan replied:
@sameetandpotatoes Change this on service side as well ^

narendasan replied:
https://github.com/acm-uiuc/groot-merch-service/blob/7a6cf5d68565595dbe540dc92a0b6ef32e950985/models/item.rb#L49

abrandemuehl committed:
Switch to Authorization Header for Authentication

sameetandpotatoes commented:
https://github.com/acm-uiuc/groot-merch-service/pull/11

Still needs to be tested

abrandemuehl commented:
@narendasan Does that fix your concern?

narendasan reviewed (approved):
Lgtm

abrandemuehl committed:
Fix up errors

abrandemuehl committed:
Make status page return useful information

The flask app is now multithreaded. Only one request will handle merch
at a time through /vend, and the /status endpoint will return 200 if the
machine is available to vend, and 503 if it is currently vending

abrandemuehl commented:
I've added the status endpoint, which will tell the service if the machine is ready to vend, as well as fixed up the errors that exist
